### PR TITLE
Fix: /tf and /static_tf topics' inconsistencies

### PR DESCRIPTION
**Pull request by Arun-Prasad-V**
Fixing #2522 

**Issues Fixed**:
1. `/static_tf` topic is getting listed even when `publish_tf:=false`
2. When the `tf_publish_rate` param is dynamically changed from 0.0 to >0.0, `/tf` topic is not getting published
3. `/tf` topic has only the TFs of last started sensor and not the TFs of previously started sensors

### CODE REVIEW

**Review by SamerKhshiboun — COMMENTED** [review]
Nice catch @Arun-Prasad-V ! Working good also on my environment.
As we fix this, I think we can add a warning also when disabling the dynamic tf
so, as we have in line 989:
`ROS_WARN("Publishing dynamic camera transforms (/tf) at %g Hz", _tf_publish_rate);`
We can add something similar at line ~1002, something like:
`ROS_WARN("Publishing dynamic camera transforms to topic (/tf) has been stopped");`

so, when user do
`ros2 param set /camera/camera tf_publish_rate 0.0`
he will get warning about that

File: realsense2_camera/src/parameters.cpp [hunk 1]
```diff
@@ -37,7 +37,6 @@ void BaseRealSenseNode::getParameters()
                 startDynamicTf();
             });
     _parameters_names.push_back(param_name);
-    startDynamicTf();
```
Comment (Nir-Az) [inline realsense2_camera/src/parameters.cpp:40]: Does this change the behavior from default start publishing to not?

Comment (Arun-Prasad-V) [reply realsense2_camera/src/parameters.cpp:40]: When the node is started, "**startDynamicTf()**" is getting called twice. So, the prints are also coming twice as shown below:
image

With this change, it will be called only once. 

Basic testing is done. Will test for corner cases as well.

**Comment by SamerKhshiboun** [comment]
As I tested it now, if we put ```pulish_tf:=false```,
and in another terminal run```ros2 topic list```, we will still see /tf_static as topic (it is an empty one, not publishing anything)
I would prefer to not starting this topic at all.
We can do this by moving `_static_tf_broadcaster` init to be inside the publishStaticTransform
It will look something like this:

```
void BaseRealSenseNode::publishStaticTransforms(std::vector<rs2::stream_profile> profiles)
{
    // Publish static transforms
    if (_publish_tf)
    {
        if(!_static_tf_broadcaster)
        {
            // intra-process do not support latched QoS, so we need to disable intra-process for this topic
            rclcpp::PublisherOptionsWithAllocator<std::allocator<void>> options;
            options.use_intra_process_comm = rclcpp::IntraProcessSetting::Disable;

#ifndef DASHING
            _static_tf_broadcaster = std::make_shared<tf2_ros::StaticTransformBroadcaster>(_node, tf2_ros::StaticBroadcasterQoS(), std::move(options));
#else
            _static_tf_broadcaster = std::make_shared<tf2_ros::StaticTransformBroadcaster>(_node, rclcpp::QoS(100), std::move(options));
#endif
        }
        
        for (auto &profile : profiles)
        {
            calcAndPublishStaticTransform(profile, _base_profile);
        }
.
.
.
```

This way, we will not see any of `/tf_static` and `/tf` topics if `publish_tf:=false`

File: realsense2_camera/launch/rs_launch.py [hunk 2]
```diff
@@ -58,6 +58,7 @@
                            {'name': 'initial_reset',                'default': 'false', 'description': "''"},                           
                            {'name': 'allow_no_texture_points',      'default': 'false', 'description': "''"},                           
                            {'name': 'pointcloud.ordered_pc',        'default': 'false', 'description': ''},
+                           {'name': 'publish_tf',                   'default': 'true', 'description': 'enable/disable publishing tf'},
```
Comment (SamerKhshiboun) [inline realsense2_camera/launch/rs_launch.py]: I would prefer to explain more in the description field, specially for these two parameters since they can confuse users.
- `publish_tf`: [bool] enable/disable publishing static or dynamic TF
- `tf_publish_rate`: [double] rate in HZ for publishing dynamic TF


Also, I would prefer to explanation and elaborate more in the README.md file, where we explain about `publish_rate `and `tf_publish_rate`
1- to explain that `publish_tf:=false` will not publish any static/dynamic TFs, even if `tf_publish_rate `is bigger than 0.0
2- to explain that `publish_tf:=true`, will automatically publish static TFs. If dynamic TFs are needed, user should set the `tf_publish_rate` for it, and it should be > 0.0
3- to mention that the default value is `publish_tf:=true` with tf_`publish_rate:=0.0` which means publishing static TFs as default behavior.

**Comment by SamerKhshiboun** [comment]
Take into mind that this fix is not covering the whole problem from issue #2522 
@Tristis116 rightly claims that not all transformations are transferred when publish Dynamic TFs (/tf topic), since we clear the TF messages. Reference: look for  `_static_tf_msgs.clear();`
Every time we restarts a sensor, we clear these messages. 
If we put some debug messages, it looks like something like this:

image

The static broadcaster (`tf2_ros::StaticTransformBroadcaster`), as a Latched Topic, works like it accumulates inside him all the messages. So, clear of messages is not harming it. Every new sensor we turn on during runtime, the broadcaster will add it's TF to the rest of TFs already in his hand.
Turning off Sensors, will not change anything, and will not remove any TF msg from the published info, and this is okay.

But the dynamic broadcaster (`tf2_ros::TransformBroadcaster` )is not Latched topic, and it works like normal topics. It looks every time on the _static_tf_msgs we have, and tries to broadcast it. So usually, you will see in `/tf` topic only the last started sensor TFs, and that is a wrong behavior.

We need to figure out how to solve it, maybe to use a HashMap instead of vector of tf_msgs, (see `std::vector<geometry_msgs::msg::TransformStamped> _static_tf_msgs;`)  where the KEY is the wanted profile or something like a string TF_FROM_TO, and then we clear/fill only by this key.



P.S. 
I don't know why we publish dynamic TFs at all, since all the camera parts are statically connected.
So, in our Camera modules, dynamic TFs are the same as static TFs, but they are being published in a specific rate, while the static TFs are being published only when needed (usually once..)

**Comment by Arun-Prasad-V** [comment]
Added fix for the issue "_**`/tf` topic has only the TFs of last started sensors**_"

**Root cause**:
- The TFs are stored in `static_tf_msgs` vector
- This vector is getting cleared everytime a sensor is started - that's why /tf has only TFs of last started sensor 

**Fix details**:
- Not clearing `static_tf_msgs` vector everytime
- When a sensor is started, add its corresponding TF frames to that vector
- When a sensor is stopped, erase its corresponding TF frames alone in that vector. So that the TFs of other sensors will be alive
    - The corresponding TFs to be erased are searched and identified in the vector using `frame_id` & `child_frame_id`

File: realsense2_camera/src/base_realsense_node.cpp [hunk 3]
```diff
@@ -855,6 +846,37 @@ tf2::Quaternion BaseRealSenseNode::rotationMatrixToQuaternion(const float rotati
     return tf2::Quaternion(q.x(), q.y(), q.z(), q.w());
 }
 
+void BaseRealSenseNode::unpublish_static_tf(const std::string& frame_id,
+                                            const std::string& child_frame_id)
+{
+    std::vector<geometry_msgs::msg::TransformStamped>::iterator it;
+    struct find_tf
```
Comment (SamerKhshiboun) [inline realsense2_camera/src/base_realsense_node.cpp]: I think it is better to move this struct outside this method.
and maybe name it with another name since using find_tf with find_if might be confusing while reading the code.

File: realsense2_camera/src/rs_node_setup.cpp [hunk 4]
```diff
@@ -351,6 +374,11 @@ void BaseRealSenseNode::updateSensors()
                 }
             }
         }
+        if (_publish_tf)
+        {
+            std::lock_guard<std::mutex> lock_guard(_publish_tf_mutex);
+            publishStaticTransforms();
```
Comment (SamerKhshiboun) [inline realsense2_camera/src/rs_node_setup.cpp:365]: if we publish static transforms only at this point, then I think we should change `calcAndPublishStaticTransforms` name to be just `calculcateStaticTransforms`
or even `calcualteTransforms` since we use it for static and dynamic TFs
This is affecting line 358

File: realsense2_camera/src/rs_node_setup.cpp [hunk 5]
```diff
@@ -195,6 +195,26 @@ void BaseRealSenseNode::stopPublishers(const std::vector<stream_profile>& profil
         }
         _metadata_publishers.erase(sip);
         _extrinsics_publishers.erase(sip);
+
+        if (_publish_tf)
+        {
+            std::lock_guard<std::mutex> lock_guard(_publish_tf_mutex);
```
Comment (SamerKhshiboun) [inline realsense2_camera/src/rs_node_setup.cpp]: I think this part from line 201 to line 217 should be in a separate function.
Like we have `PublishStaticTransforms()` which uses `publish_static_tf` inside it, we can have `UnpublishStaticTransforms(<params>)` which will use `unpublish_static_tf` inside it.

File: realsense2_camera/src/base_realsense_node.cpp [hunk 6]
```diff
@@ -1016,12 +1052,22 @@ void BaseRealSenseNode::startDynamicTf()
             _tf_t->join();
             _tf_t.reset();
             _dynamic_tf_broadcaster.reset();
+            ROS_WARN("Stopped publishing dynamic camera transforms (/tf)");
+        }
+        else
+        {
+            ROS_WARN("Currently not publishing dynamic camera transforms (/tf)");
```
Comment (SamerKhshiboun) [inline realsense2_camera/src/base_realsense_node.cpp]: Just a question, can you remind me when we will get to here ? In which scenario ?

File: realsense2_camera/src/base_realsense_node.cpp [hunk 7]
```diff
@@ -981,28 +1003,42 @@ void BaseRealSenseNode::SetBaseStream()
...
-        for (auto &profile : profiles)
-        {
-            calcAndPublishStaticTransform(profile, _base_profile);
-        }
-        if (_static_tf_broadcaster)
-            _static_tf_broadcaster->sendTransform(_static_tf_msgs);
+        _static_tf_broadcaster.reset();
     }
+
+    // intra-process do not support latched QoS, so we need to disable intra-process for this topic
```
Comment (SamerKhshiboun) [inline realsense2_camera/src/base_realsense_node.cpp]: I would move this part from line 1008 to line 1024 into a seperate function
`resetStaticTransformBroadcaster(_static_tf_broadcaster)`

**Comment by SamerKhshiboun** [comment]
Great job @Arun-Prasad-V.
I think now we cover all flows and it is working good.
I gave you some comments, and I think we can move all `base_realsense::<TFs functions>` to a new cpp file `tfs.cpp` to be dedicated to all TF methods (no need for header file. see `parameters.cpp` for example).
If you do that, remember to add this cpp to the cmake file

**Review by SamerKhshiboun — APPROVED** [review]
LGTM
